BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c

```diff
@@ -337,6 +429,71 @@ LocateComponentEntry (
   return EFI_SUCCESS;
 }
 
+/**
+  Get Next Component in a container.
+
+  @param[in]       ContainerSig    Container signature or component type.
+  @param[in, out]  ComponentName   Pointer to receive next component name.
+
+  @retval EFI_INVALID_PARAMETER    ComponentName is NULL.
+  @retval EFI_NOT_FOUND            Cannot find next component.
+  @retval EFI_SUCCESS              Next component was found successfully.
+
+**/
+EFI_STATUS
+EFIAPI
+GetNextComponent (
+  IN     UINT32     ContainerSig,
+  IN     UINT32    *ComponentName
+)
+{
+  EFI_STATUS         Status;
+  CONTAINER_HDR     *ContainerHdr;
+  CONTAINER_ENTRY   *ContainerEntry;
+  COMPONENT_ENTRY   *CurrEntry;
+  COMPONENT_ENTRY   *NextEntry;
+  UINT32             Index;
+
+  if (ComponentName == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Search container header from cache
+  Status = EFI_NOT_FOUND;
+  ContainerEntry = GetContainerBySignature (ContainerSig);
+  if (ContainerEntry == NULL) {
+    return Status;
+  }
+
+  ContainerHdr = (CONTAINER_HDR *)ContainerEntry->HeaderCache;
+  if (ContainerHdr->Count == 0) {
+    return Status;
+  }
+
+  CurrEntry = (COMPONENT_ENTRY *)&ContainerHdr[1];
+  for (Index = 0; Index < (UINT32)(ContainerHdr->Count - 1); Index++) {
+    NextEntry = (COMPONENT_ENTRY *)((UINT8 *)(CurrEntry + 1) + CurrEntry->HashSize);
+    if ((CurrEntry->Attribute & COMPONENT_ENTRY_ATTR_RESERVED) == 0) {
+      if (*ComponentName == 0) {
+        *ComponentName = CurrEntry->Name;
+        Status = EFI_SUCCESS;
+        break;
+      } else if (*ComponentName == CurrEntry->Name) {
+        if ((NextEntry->Attribute & COMPONENT_ENTRY_ATTR_RESERVED) == 0) {
+          *ComponentName = NextEntry->Name;
+          Status = EFI_SUCCESS;
+          break;
+        } else {
+          Status = EFI_NOT_FOUND;
```

Comment:
if next entry is reserved,  shouldn't we continue to find next normal entry and return it ?

Reply:
I am assuming mono sig component or all reserved components to be at the end of the container. If not, I can check for count and continue to search for all other unreserved components.

Reply:
it should be that way.  but for code flow, let us do not assume that.  Let us clean the logic:
1. Find first non-reserved entry.
2. If  *Component=0, return this entry.
3. If  *Component = Entry->Name,  return next non-reserved entry.

Reply:
The API name - GetNextComponent() is a little confusing. It looks like just an iteration. So, GetNextAvailableComponent() would be better?